From a2da104facddc649f5ac477dd7125c468d4203c6 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Mon, 27 Oct 2025 16:23:47 +0100
Subject: [PATCH 1004/1013] igb: assume no oob traffic on oob mode switch

We may assume that no RX/TX oob traffic can happen while the {disable,
enable}_oob handlers are running. This MUST be enforced by the oob
netstack.

For this reason, synchronizing with the NAPI and/or oob TX is
pointless (the former may even cause the operation to hang
indefinitely).

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 drivers/net/ethernet/intel/igb/igb_main.c | 28 +----------------------
 1 file changed, 1 insertion(+), 27 deletions(-)

diff --git a/drivers/net/ethernet/intel/igb/igb_main.c b/drivers/net/ethernet/intel/igb/igb_main.c
index e7f35325b362..22bb1bde10dd 100644
--- a/drivers/net/ethernet/intel/igb/igb_main.c
+++ b/drivers/net/ethernet/intel/igb/igb_main.c
@@ -758,20 +758,13 @@ static int igb_enable_oob(struct net_device *netdev)
 		if (!adapter->rx_ring[n]->rx_oob_pool)
 			return -ENOMEM;
 
-	for (n = 0; n < adapter->num_q_vectors; n++) {
-		struct igb_q_vector *q_vector = adapter->q_vector[n];
-		napi_disable(&q_vector->napi);
-	}
-
-	netif_tx_lock_bh(netdev);
-
 	if (adapter->flags & IGB_FLAG_HAS_MSIX) {
 		for (n = 0; n <= adapter->num_q_vectors && n <= MAX_Q_VECTORS; n++) {
 			ret = irq_switch_oob(adapter->msix_entries[n].vector, true);
 			if (ret) {
 				while (--n >= 0)
 					irq_switch_oob(adapter->msix_entries[n].vector, false);
-				goto out;
+				return ret;
 			}
 		}
 		pr_info("%s: enabled out-of-band I/O mode (MSI-X)\n", netdev_name(netdev));
@@ -782,13 +775,6 @@ static int igb_enable_oob(struct net_device *netdev)
 				netdev_name(netdev),
 				adapter->flags & IGB_FLAG_HAS_MSI ? "MSI" : "legacy-int");
 	}
-out:
-	netif_tx_unlock_bh(netdev);
-
-	for (n = 0; n < adapter->num_q_vectors; n++) {
-		struct igb_q_vector *q_vector = adapter->q_vector[n];
-		napi_enable(&q_vector->napi);
-	}
 
 	return ret;
 }
@@ -798,12 +784,6 @@ static void igb_disable_oob(struct net_device *netdev)
 	struct igb_adapter *adapter = netdev_priv(netdev);
 	int n;
 
-	for (n = 0; n < adapter->num_q_vectors; n++) {
-		struct igb_q_vector *q_vector = adapter->q_vector[n];
-		napi_disable(&q_vector->napi);
-		netif_tx_lock_bh(netdev);
-	}
-
 	if (adapter->flags & IGB_FLAG_HAS_MSIX) {
 		for (n = 0; n <= adapter->num_q_vectors && n <= MAX_Q_VECTORS; n++)
 			irq_switch_oob(adapter->msix_entries[n].vector, false);
@@ -814,12 +794,6 @@ static void igb_disable_oob(struct net_device *netdev)
 			netdev_name(netdev),
 			adapter->flags & IGB_FLAG_HAS_MSI ? "MSI" : "legacy-int");
 	}
-
-	for (n = 0; n < adapter->num_q_vectors; n++) {
-		struct igb_q_vector *q_vector = adapter->q_vector[n];
-		netif_tx_unlock_bh(netdev);
-		napi_enable(&q_vector->napi);
-	}
 }
 
 static void igb_create_oob_pool(struct igb_ring *ring)
-- 
2.30.2

